### ISSUE TITLE
Reload templates without restarting

### ISSUE THREAD

**Issue by mastaofdisasta** [FRosner/cluster-broccoli#229]
Cluster broccoli could watch for changes in template directory (inotify?) or provide some HTTP API endpoint to request template refresh.

**Comment by FRosner** [comment]
Thanks for creating that issue @mastaofdisasta! I think a good first step is to support template reloading through a signal in case people are deploying the Broccoli templates through the template stanza from Nomad. What do you think?

We could use `SIGUSR1` to reload the templates?

Providing an HTTP endpoint to reload is also possible but we would have to add some authorization around that. I don't see a huge benefit over signals, esp. as reloading only makes sense after changes to the files have been made. Instead, we should allow changing templates through the API / Web UI so that administrators can create new templates during runtime without access to the file system.

PS: Sorry for the late response I was not in the country.

**Comment by swsnr** [comment]
I suggest tearing the `TemplateService` apart and move template loading into the following structure of "layered template sources":

```scala

trait TemplatesSource {
  def loadTemplates(): Seq[Template]
}

class DirectoryTemplatesSources(directory: String) extends TemplatesSource {
  def loadTemplates(): Seq[Template] = ???
}

@Singleton
class CachedTemplatesSource(source: TemplatesSource) extends TemplatesSource {
  @volatile private var templatesCache: Option[Seq[Template]] = ???

  override def loadTemplates(): Seq[Template] =
    templatesCache match {
      case Some(templates) => templates
      case None => {
        refresh()
        templatesCache.get
      }
    }

  def refresh(): Unit = templatesCache = Some(source.loadTemplates())
}

class CacheRefresher(source: CachedTemplatesSource) extends TemplatesSource {

  def wheneverUSR1(): Unit = {
    source.refresh()
  }

  override def loadTemplates(): Seq[Template] = source.loadTemplates()
}
```

Each part represents a single concern now:

1. Loading templates from disk
2. Maintaining an in-memory cache of templates
3. Refreshing the cache on SIGURS1

The TemplateService would get a TemplatesSource injected via DI; a new "TemplatesModule" would configure the appropriate TemplatesSource for injection. /cc @Gerrrr

**Comment by FRosner** [comment]
Is everything done here @Gerrrr ?

**Comment by Gerrrr** [comment]
Not yet.

I still have to mention in the release notes for 0.7.0 that the default paths for instance and template folders changed from `/instance` and `/templates` to `/cluster-broccoli-dist/instance` and `/cluster-broccoli-dist/templates` respectively.

**Comment by FRosner** [comment]
I see a possible conflict with this one? https://github.com/FRosner/cluster-broccoli/pull/281

**Comment by swsnr** [comment]
@FRosner @Gerrrr I thought these changes happened in #265?

In this case I'd merge #265 first and then update #281 afterwards before merging it.

**Comment by Gerrrr** [comment]
These changes happened in #255 that is already merged.

**Comment by swsnr** [comment]
Ah, I wasn't aware.  I'll update #281 accordingly :)

**Comment by swsnr** [comment]
@Gerrrr It seems though that these changes only made it into `docker/test/Dockerfile`, but not `docker/prod/Dockerfile`?

**Comment by Gerrrr** [comment]
@lunaryorn yes you are right. It should be the same in both files. Do you want me to include the fix in #265?

**Comment by swsnr** [comment]
@Gerrrr No, I've redone the production image build in #281 which fixes this as well.
